### ISSUE TITLE
Regression: `alt_complete!` fails to properly infer type from pattern mapping

### ISSUE THREAD

**Issue by yodaldevoid** [Geal/nom#511]
The below code fails to compile with the error below that. Interesting to note is that changing `alt_complete!` to `alt!` fixes this error as does using explicit `map!`s.

```Rust
#[derive(PartialEq, Debug)]
enum OprPrefix {
    Negate,
    BitNot,
    Not,
}

named!(opr_prefix<OprPrefix>, alt_complete!(
    tag!("!") => { |_| OprPrefix::Not } |
    tag!("~") => { |_| OprPrefix::BitNot } |
    tag!("-") => { |_| OprPrefix::Negate }
));
```
```
error[E0308]: mismatched types
   --> src/parser.rs:139:1
    |
139 |   named!(opr_prefix<OprPrefix>, alt_complete!(
    |  _^ starting here...
140 | |     tag!("!") => { |_| OprPrefix::Not } |
141 | |     tag!("~") => { |_| OprPrefix::BitNot } |
142 | |     tag!("-") => { |_| OprPrefix::Negate }
143 | | ));
    | |___^ ...ending here: expected &[u8], found enum `parser::OprPrefix`
    |
    = note: expected type `&nom::IResult<&[u8], &[u8]>`
               found type `&nom::IResult<&[u8], parser::OprPrefix>`
    = note: this error originates in a macro outside of the current crate
```

**Comment by Geal** [comment]
thanks for the report. This one should be easy to fix, I'll take care of it tomorrow or tuesday

**Comment by Geal** [comment]
This should be fixed with 7fd15580e30

**Comment by ekse** [comment]
I just hit this bug too when updating to nom 3.0, 7fd1558 did fix it for me.

**Comment by ekse** [comment]
Hi @Geal, could you do a new release on crates.io? Thanks.

**Comment by Geal** [comment]
I'll do a release on saturday, I have a fe things to address before

**Comment by Geal** [comment]
hi! This is available in nom 3.1

**Comment by ekse** [comment]
Thanks! I updated my project to version 3.1.
